### PR TITLE
Add missing React-related eslint rules

### PR DESCRIPTION
**Pull request by odididi**

### CODE REVIEW

File: jsx.js [hunk 1]
```diff
@@ -13,7 +13,78 @@ module.exports = {
...
+    'react/jsx-curly-spacing': [2, 'never'],
+    'react/jsx-equals-spacing': [0],
+    'react/jsx-filename-extension': [0],
+    'react/jsx-first-prop-new-line': [2],
+    'react/jsx-fragments': [2, 'syntax'],
+    'react/jsx-handler-names': [0],
+    'react/jsx-indent': [2, 2, {checkAttributes: true, indentLogicalExpressions: true}],
+    'react/jsx-indent-props': [2, 2],
+    'react/jsx-max-depth': [0],
+    'react/jsx-max-props-per-line': [2],
```
Comment (d-eisenga) [inline jsx.js:41]: I'd prefer to leave this disabled, I commonly put multiple props on a single line for e.g. SVG elements
(like `<circle cx={50} cy={50} r={25}/>`), and files can get pretty long this way, often with just a handful of characters on many of its lines.

File: jsx.js [hunk 2]
```diff
@@ -13,7 +13,78 @@ module.exports = {
...
+    'react/jsx-newline': [2, {prevent: true}],
+    'react/jsx-no-constructed-context-values': [2],
+    'react/jsx-no-literals': [0],
+    'react/jsx-no-script-url': [0],
+    'react/jsx-no-useless-fragment': [2, {allowExpressions: true}],
+    'react/jsx-one-expression-per-line': [2, {allow: 'literal'}],
```
Comment (d-eisenga) [inline jsx.js:47]: I'd prefer `{allow: 'single-child'}` as configuration for this rule, again not to have a lot of very short lines.

File: jsx.js [hunk 3]
```diff
@@ -13,7 +13,78 @@ module.exports = {
...
+      assignment: 'parens-new-line',
+      return: 'parens-new-line',
+      arrow: 'parens-new-line',
+      condition: 'parens-new-line',
+      logical: 'parens-new-line',
+      prop: 'parens-new-line'
+    }],
+    'react/no-access-state-in-setstate': [2],
+    'react/no-adjacent-inline-elements': [0],
+    'react/no-array-index-key': [2],
```
Comment (d-eisenga) [inline jsx.js:65]: This is a good rule in principle, but often we just don't have anything better to index on, and I suspect we'd have to disable this rule more often than not.

Comment (odididi) [reply jsx.js:65]: could we add it as a warning perhaps?

File: jsx.js (hunk 2 above)
Comment (odididi) [reply jsx.js:47]: 👍

File: jsx.js (hunk 1 above)
Comment (odididi) [reply jsx.js:41]: i agree about such edge cases, but i feel there might be more less readable instances if we disable it

File: jsx.js (hunk 3 above)
Comment (Avaq) [reply jsx.js:65]: > often we just don't have anything better to index on

I feel like it's quite rare to display items from an array where no combination of a subset of properties from the array item uniquely identifies it. I've encountered many usages of keying by the array index where it seems like the developer just didn't put enough thought into what the identifying properties of the array items are.

In cases where it's really not possible to uniquely identify array items using a subset of their properties, I find that it is indicative of a common anti-pattern, where the developer should just have used plain JSX, but "abstracted away" the duplicate code by using an array.

Comment (Avaq) [reply jsx.js:65]: To be clear: I'm in favour of making this rule an error.

Comment (Avaq) [reply jsx.js:65]: > could we add it as a warning perhaps?

The only rules we should make warnings of are things that we are happy to be piling up on `main`. Anything else should be an error or off.

File: jsx.js (hunk 1 above)
Comment (Avaq) [reply jsx.js:41]: Wait, you want all JSX tags to be spread over rmultiple lines?

So you'd prefer seeing:

```jsx
<div onClick={onClick} style={styles.div}>
  <a href="google.com" target="_blank">
    <Icon name="google" type="filled" />
  </a>
</div>
```

like this:

```jsx
<div
  onClick={onClick}
  style={styles.div}
>
  <a
    href="google.com"
    target="_blank"
  >
    <Icon
      name="google"
      type="filled"
    />
  </a>
</div>
```

I'd be strongly opposed.

**Comment by Avaq** [comment]
Before we merge #26 and #25 into `main`, I'll make a temporary `next`-branch and merge them both there, so we can try it out with `npm i wearereasonablepeople/eslint-config-warp#next`.

File: jsx.js (hunk 1 above)
Comment (odididi) [reply jsx.js:41]: I would indeed prefer the first, so i rest my case

Comment (odididi) [reply jsx.js:41]: Is there value on specifying the maximum number of props, or just disable it?

Comment (Avaq) [reply jsx.js:41]: I think we should disable it. Line length is good enough to dictate the max props, I think.

**Review by Avaq — APPROVED** [review]

**Pull request merged**